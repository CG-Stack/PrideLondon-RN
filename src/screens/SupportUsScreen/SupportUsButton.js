// @flow
import React from "react";
import { Image, Linking, PixelRatio, StyleSheet, View } from "react-native";
import type { StyleObj } from "react-native/Libraries/StyleSheet/StyleSheetTypes";
import type { NavigationScreenProp, NavigationState } from "react-navigation";
import chevronRightCircleWhite from "../../../assets/images/chevronRightCircleWhite.png";
import externalLinkBlue from "../../../assets/images/externalLinkBlue.png";
import externalLinkWhite from "../../../assets/images/externalLinkWhite.png";
import Text from "../../components/Text";
import Touchable from "../../components/Touchable";
import {
  lightNavyBlueColor,
  whiteColor,
  supportUsButtonShadow
} from "../../constants/colors";

type Props = {
  color: string,
  title: string,
  description: string,
  navigation: NavigationScreenProp<NavigationState>,
  url: string,
  isExternalLink?: boolean,
  contrast?: boolean,
  style?: StyleObj
};

const pickLinkIcon = (isExternalLink: boolean, contrast: boolean) => {
  if (isExternalLink) {
    return contrast ? externalLinkBlue : externalLinkWhite;
  }

  return chevronRightCircleWhite;
};

class SupportUsButton extends React.PureComponent<Props> {
  onPress = () => {
    if (this.props.isExternalLink) {
      Linking.openURL(this.props.url);
    } else {
      this.props.navigation.navigate(this.props.url);
    }
  };

  render() {
    const {
      color,
      title,
      description,
      isExternalLink,
      contrast,
      style
    } = this.props;

    return (
      <Touchable
        onPress={this.onPress}
        style={[styles.button, { backgroundColor: color }, style]}
      >
        <View style={styles.titleContainer}>
          <Text type="h1" style={contrast ? styles.textContrast : styles.text}>
            {title}
          </Text>
          <Image
            source={pickLinkIcon(!!isExternalLink, !!contrast)}
            style={styles.titleIcon}
          />
        </View>
        <Text type="h3" style={contrast ? styles.textContrast : styles.text}>
          {description}
        </Text>
      </Touchable>
    );
  }
}

const styles = StyleSheet.create({
  button: {
    justifyContent: "center",
    height: 100 * PixelRatio.getFontScale(),
    borderRadius: 5,
    paddingHorizontal: 28,

    // The below properties are required for iOS shadow
    shadowColor: supportUsButtonShadow,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 1,
    shadowRadius: 3,

    // The below properties are required for Android shadow
    borderWidth: 0,
    elevation: 3
  },
  titleContainer: {
    flexDirection: "row"
  },
  titleIcon: {
    marginLeft: 8,
    alignSelf: "center",
    transform: [{ translateY: -4 * PixelRatio.getFontScale() }]
  },
  text: {
    color: whiteColor
  },
  textContrast: {
    color: lightNavyBlueColor
  }
});

export default SupportUsButton;
