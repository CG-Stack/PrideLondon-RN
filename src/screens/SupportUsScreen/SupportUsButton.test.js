// @flow
import React from "react";
import { Linking } from "react-native";
import { shallow } from "enzyme";
import SupportUsButton from "./SupportUsButton";

let openURLSpy;
beforeEach(() => {
  openURLSpy = jest
    .spyOn(Linking, "openURL")
    .mockImplementation(() => Promise.resolve());
});

it("renders correctly", () => {
  const output = shallow(
    <SupportUsButton
      color="red"
      title="Fruit..."
      description="Apples, Bananas, Tomatos"
      navigation={null}
      url="https://example.com"
    />
  );
  expect(output).toMatchSnapshot();
});

it("picks different icon for external links", () => {
  const output = shallow(
    <SupportUsButton
      color="red"
      title="Fruit..."
      description="Apples, Bananas, Tomatos"
      navigation={null}
      url="https://example.com"
      isExternalLink
    />
  );
  expect(output).toMatchSnapshot();
});

it("picks different icon for external links with contrast color", () => {
  const output = shallow(
    <SupportUsButton
      color="red"
      title="Fruit..."
      description="Apples, Bananas, Tomatos"
      navigation={null}
      url="https://example.com"
      isExternalLink
      contrast
    />
  );
  expect(output).toMatchSnapshot();
});

it("navigates to route onPress", () => {
  const navigation = {
    navigate: jest.fn()
  };
  const output = shallow(
    <SupportUsButton
      color="red"
      title="Fruit..."
      description="Apples, Bananas, Tomatos"
      navigation={navigation}
      url="DONATE"
    />
  );
  output.simulate("press");

  expect(navigation.navigate).toHaveBeenCalledWith("DONATE");
});

it("navigates to external link onPress", () => {
  const output = shallow(
    <SupportUsButton
      color="red"
      title="Fruit..."
      description="Apples, Bananas, Tomatos"
      navigation={null}
      url="https://example.com"
      isExternalLink
    />
  );
  output.simulate("press");

  expect(openURLSpy).toHaveBeenCalledWith("https://example.com");
});

afterEach(() => {
  openURLSpy.mockRestore();
});
