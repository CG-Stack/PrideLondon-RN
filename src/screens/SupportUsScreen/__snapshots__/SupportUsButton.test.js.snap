// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`picks different icon for external links 1`] = `
<Touchable
  accessibilityComponentType="button"
  accessibilityTraits={
    Array [
      "button",
    ]
  }
  delayPressIn={50}
  onPress={[Function]}
  style={
    Array [
      Object {
        "borderRadius": 5,
        "borderWidth": 0,
        "elevation": 3,
        "height": 200,
        "justifyContent": "center",
        "paddingHorizontal": 28,
        "shadowColor": "rgba(0, 0, 0, 0.2)",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 3,
      },
      Object {
        "backgroundColor": "red",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      markdown={false}
      style={
        Object {
          "color": "#ffffff",
        }
      }
      type="h1"
    >
      Fruit...
    </Text>
    <Image
      source={1}
      style={
        Object {
          "alignSelf": "center",
          "marginLeft": 8,
          "transform": Array [
            Object {
              "translateY": -8,
            },
          ],
        }
      }
    />
  </View>
  <Text
    markdown={false}
    style={
      Object {
        "color": "#ffffff",
      }
    }
    type="h3"
  >
    Apples, Bananas, Tomatos
  </Text>
</Touchable>
`;

exports[`picks different icon for external links with contrast color 1`] = `
<Touchable
  accessibilityComponentType="button"
  accessibilityTraits={
    Array [
      "button",
    ]
  }
  delayPressIn={50}
  onPress={[Function]}
  style={
    Array [
      Object {
        "borderRadius": 5,
        "borderWidth": 0,
        "elevation": 3,
        "height": 200,
        "justifyContent": "center",
        "paddingHorizontal": 28,
        "shadowColor": "rgba(0, 0, 0, 0.2)",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 3,
      },
      Object {
        "backgroundColor": "red",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      markdown={false}
      style={
        Object {
          "color": "#2d2f7f",
        }
      }
      type="h1"
    >
      Fruit...
    </Text>
    <Image
      source={1}
      style={
        Object {
          "alignSelf": "center",
          "marginLeft": 8,
          "transform": Array [
            Object {
              "translateY": -8,
            },
          ],
        }
      }
    />
  </View>
  <Text
    markdown={false}
    style={
      Object {
        "color": "#2d2f7f",
      }
    }
    type="h3"
  >
    Apples, Bananas, Tomatos
  </Text>
</Touchable>
`;

exports[`renders correctly 1`] = `
<Touchable
  accessibilityComponentType="button"
  accessibilityTraits={
    Array [
      "button",
    ]
  }
  delayPressIn={50}
  onPress={[Function]}
  style={
    Array [
      Object {
        "borderRadius": 5,
        "borderWidth": 0,
        "elevation": 3,
        "height": 200,
        "justifyContent": "center",
        "paddingHorizontal": 28,
        "shadowColor": "rgba(0, 0, 0, 0.2)",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 3,
      },
      Object {
        "backgroundColor": "red",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      markdown={false}
      style={
        Object {
          "color": "#ffffff",
        }
      }
      type="h1"
    >
      Fruit...
    </Text>
    <Image
      source={1}
      style={
        Object {
          "alignSelf": "center",
          "marginLeft": 8,
          "transform": Array [
            Object {
              "translateY": -8,
            },
          ],
        }
      }
    />
  </View>
  <Text
    markdown={false}
    style={
      Object {
        "color": "#ffffff",
      }
    }
    type="h3"
  >
    Apples, Bananas, Tomatos
  </Text>
</Touchable>
`;
